Fix sampling of pm.Deterministic and add auto_deterministics to identify imputed deterministics

ricardoV94 opened the pull request:
@brandonwillard It seems there were still some issues with missing values. 

```python
import pymc3 as pm; import numpy as np 
with pm.Model() as m: 
     x = pm.Gamma('x', 1, 1, size=4, observed=np.array([1, 1, 1, np.nan]))
     trace = pm.sample()   

# Observed values are being resampled during pm.sample()
# which looks buggy even though it doesn't seem to impact the logp
# It definitely looks strange from a user standpoint
trace['x'].std(0)
# array([0.96463621, 1.02997068, 1.01582103, 1.00555309])

# In addition values from `x_missing` and `x` do not match.
np.mean(trace['x'][:, -1] == trace['x_missing'].squeeze())
# 0.0
```

This PR does not fix this, but simply hides `x` from the normal trace, by adding them to a new category of  `partobs_deterministics` variables in the model object. This also makes it easier to provide the same standard output that was given in V3, see below: 


This table shows the outputs a user gets in `V3` in the traces from prior/predictive sampling and normal sampling when there are missing variables present:

function | y | y_miss | y_obs | deterministics
--- | --- | --- | --- | ---
prior_pred | ✓ | ✓ | NA  | ✓ 
sample | x | ✓ | NA  | ✓ 
posterior_pred | ✓ | x | NA | x


In the current `V4` branch:

function | y | y_miss | y_obs | deterministics
--- | --- | --- | --- | ---
prior_pred | ✓ | ✓ | ✓  | ✓ 
sample | **!** | ✓ | x  | ✓ 
posterior_pred | x | x | ✓ | x

Importantly:
1. `y` currently appears in the trace from `pm.sample()` with random numbers everywere (observed and missing), as discussed at the top
2. `y` is missing from posterior predictive (only `y_obs` appears)

After this PR:
function | y | y_miss | y_obs | deterministics
--- | --- | --- | --- | ---
prior_pred | ✓ | ✓ | ✓ | ✓ 
sample | x | ✓ | x |  ✓ 
posterior_pred | ✓ | x | ✓ | x

Also closes https://github.com/pymc-devs/pymc3/issues/4488

***

Depending on what your PR does, here are a few things you might want to address in the description:
+ [ ] what are the (breaking) changes that this PR makes?
+ [ ] important background, or details about the implementation
+ [ ] are the changes—especially new features—covered by tests and docstrings?
+ [ ] [linting/style checks have been run](https://github.com/pymc-devs/pymc3/wiki/PyMC3-Python-Code-Style)
+ [ ] [consider adding/updating relevant example notebooks](https://github.com/pymc-devs/pymc-examples)
+ [ ] right before it's ready to merge, mention the PR in the RELEASE-NOTES.md

brandonwillard commented:
> ```python
> Observed values are being resampled during pm.sample()
> # which looks buggy even though it doesn't seem to impact the logp
> # It definitely looks strange from a user standpoint
> ```

What you've described doesn't sound like a bug at all; the returned `x` is the **random variable** corresponding to the observation distribution, and, if/when it's sampled, it absolutely should produce values other than the observed ones&mdash;unless the distribution is fixed at those points, of course.

It sounds like you're saying that the results for `x` **in the trace** produced by `pm.sample` are somehow confusing, but those are simply (posterior) predictive values&mdash;like one would get from `pm.sample_posterior_predictive`.


> ```python
> # In addition values from `x_missing` and `x` do not match.
> ```

This apparent discrepancy arises for exactly the same reasons as above: the `x` values are predictive draws that have nothing to do with the `x_missing` estimated via NUTS.  The `x` values in the trace are sampled straight from the `RandomVariable`s corresponding to `x_observed` and `x_missing`, and not the `x_observed` values used in the likelihood (i.e. the actual observed values) or the `x_missing` sampled by NUTS.  If we wanted to return _that_ value of `x`, we would need to convert `x` to its log-likelihood space equivalent (e.g. using `pymc3.aesaraf.rvs_to_value_vars`).

The answer to all this probably is as simple as removing `x` from the `pm.sample` results.  We don't want to remove `x` from the likelihood(s), because that would interfere with models that use the partially observed `x`, but we definitely don't want/need to draw those values and return them in the trace.

brandonwillard reviewed (changes requested):
Regarding this fix, we don't need to create another set of variables to do this.  We only need to remove the `x` values from the result or convert it to its likelihood-space representation so that it uses the observed values and sampler estimated missing values.

ricardoV94 commented:
> The answer to all this probably is as simple as removing x from the pm.sample results. We don't want to remove x from the likelihood(s), because that would interfere with models that use the partially observed x, but we definitely don't want/need to draw those values and return them in the trace.

This is not really removing x from the likelihood graph (we have tests that would fail if that was the case). The variables that make use of x still include it in their logp.

This fix makes it much simpler, however, to identify complete observeds (the x) to include in the post predictive sampling. Otherwise I would have to do some pattern matching with the variable names or check graph dependencies between each observed and each deterministic variable, wich I couldn't do in an unmessy way.

ricardoV94 commented:
Okay, so this is also indicative that `Deterministics` are not making use of value RVs during sampling.

```python
import pymc3 as pm
with pm.Model() as m:
    x = pm.Normal('x', 0, 1)
    det_x = pm.Deterministic('det_x', x)
    y = pm.Normal('y', mu=x, sd=1, observed=[100])
    
with m:
    trace = pm.sample()

print(trace['x'].mean())  # 50
print(trace['det_x'].mean()) # -0.04
```

ricardoV94 commented:
I pushed some changes that seem to fix the Deterministics not having the RVs replaced by the value_vars during sampling. Can you check if it makes sense @brandonwillard ?

ricardoV94 commented:
So altogether this PR adds a new property to model: `unobserved_value_vars` that includes `value_vars` + `deterministics` both of which have the rvs replaced from their graphs.

This is used to fix the stochastic deterministic during normal sampling (and error when using `find_MAP()`). It seems like a good helper attribute for other (logp) sampling methods as well. 

In addition, there is a new list of `auto_deterministics` that is used to tag the automatic deterministic that is created when missing values are present (`x_observed` + `x_missing`). This makes it easy to exclude it from normal sampling (to reduce memory footprint) but include in prior / posterior predictive sampling. Users can still save it explicitly during sampling by wrapping it in a pm.Deterministic such as:

```python
with pm.Model() as m:
  x = pm.Uniform('x', 0, 1, observed=np.array([1, 1, np.nan]))
  x_det = pm.Deterministic('x_det', x)
```
Also there are tests in place that show that not including the `x` auto_deterministic in `unobserved_RVs` is not screwing up the logp graph if other variables make use of it (it simply is not being stored explicitly in the trace, but it's still there).

The behavior of sample and predictive sampling is now similar to `V3`, when there are missing values, except for the new category of `x_observed` which was added before.

ricardoV94 committed:
Add unobserved_value_vars property to Model

Fixes pm.Deterministics still using rvs during pm.Sample()
Fixes find_MAP() failing when pm.Deterministics were present in model

ricardoV94 committed:
Add auto_deterministics list to Model

Ensures that when missing variables are present in the model, the automatic deterministic (x_observed + x_missing) only appears in predictive sampling and not normal sampling.
Fixes `x` missing from prior_predictive when missing values were present (only `x_missing` was present)

brandonwillard reviewed (approved):
Great work!
